docs_nnx/guides/performance.md

```diff
@@ -0,0 +1,110 @@
+
+
+## Asynchronous dispatch
+In [benchmarks/nnx_simple_training.py](https://github.com/google/flax/blob/main/benchmarks/nnx_simple_training.py) we are increasing the layer width (features per layer) and measuring the total training time for the same model trained both with `nnx.jit` and `jax.jit`. As you can see in the graph below, after a certain model size the time spent in the traversal is completely absorbed by async dispatch. This happens when Python is able to finish the current for loop step, and reach the next `train_step` and JAX is still not done with the previous `train_step`. 
+
+![performance-graph](images/performance-graph.png)
+
+This means that you only need to worry about the `nnx.jit` overhead for small models. If you are working with a small model, check out the next section to see how you can remove the overhead.
+
+## Lowering the Python Overhead
+To remove the python overhead you can use regular `jax.jit` in combination with `nnx.split` and `nnx.merge` to stage out the traversal logic. To learn how to do this, lets first create this simple model:
+
+```{code-cell}
+from flax import nnx
+import jax
+import jax.numpy as jnp
+import optax
+
+class Model(nnx.Module):
+  def __init__(self, din, dmid, dout, rngs: nnx.Rngs):
+    self.linear = nnx.Linear(din, dmid, rngs=rngs)
+    self.bn = nnx.BatchNorm(dmid, rngs=rngs)
+    self.dropout = nnx.Dropout(0.2, rngs=rngs)
+    self.linear_out = nnx.Linear(dmid, dout, rngs=rngs)
+
+  def __call__(self, x):
+    x = nnx.relu(self.dropout(self.bn(self.linear(x))))
+    return self.linear_out(x)
+```
+
+Lets say we have this `train_step` function that is using `nnx.jit` and takes in a `model`, `optimizer`, and `metrics`, all of which are Flax NNX objects:
+
+```{code-cell}
+model = Model(2, 64, 3, rngs=nnx.Rngs(0))  # eager initialization
+optimizer = nnx.Optimizer(model, optax.adam(1e-3))  # reference sharing
+metrics = nnx.MultiMetric(
+  loss=nnx.metrics.Average('loss'),
+)
+
+@nnx.jit  # <== currently slow
+def train_step(model, optimizer, metrics, x, y):
+  def loss_fn(model):
+    y_pred = model(x)  # call methods directly
+    return ((y_pred - y) ** 2).mean()
+
+  loss, grads = nnx.value_and_grad(loss_fn)(model)
+  optimizer.update(grads)  # in-place updates
+  metrics.update(loss=loss)
+
+  return loss
+  
+for _ in range(10):
+  x, y = jnp.ones((32, 2)), jnp.zeros((32, 3))
+  loss = train_step(model, optimizer, metrics, x, y)
+```
+
+To speed it up we can use `nnx.split` before starting the training loop to create a `graphdef` and `state` pytrees for the Flax NNX objects as a group since `graphdef` and `state` are fast to traverse. Then at the beggining and end of a `jax.jit`-decorated function we can call `nnx.merge` and `nnx.split` to switch back and forth between the object and pytree representations. The important thing here is that `split` and `merge` will only run once during tracing.
```

Comment:
I think we should add a line here to explicitly say that you need to run nnx.merge and nnx.split on all the NNX objects that are part of the train step inputs.